u/melissagraham/2025-02-28/fix-13a

MelissaGraham opened the pull request:
Switch to cutouts-in-memory as the default way to retrieve cutouts. Still demonstrate how to save-and-plot once.

athornton reviewed (changes requested):
If you're putting it in `/scratch` it should go under `/scratch/$USER` because scratch is shared.  (well, ok, `f"/scratch/${os.getenv('USER')"`)  Otherwise if you have multiple concurrent users they're going to overwrite each others' files.

However...all this particular PR seems to do is `import getpass` and then not use it, so I'm very confused.

athornton commented:
Also in general if the contents are small you should probably use `tempfile.TemporaryDirectory` instead of creating your own explicitly, but since that will come out of the system temp space which may be backed by disk or may be backed by ephemeral storage (and thus come out of the host memory) that's not safe for large contents and a lot of simultaneous users, as we found out with remote Butler a few months back.

MelissaGraham commented:
@athornton  thanks very much for having a look.

The NB does use `getpass`, and does use `/scratch/<username>/temp`. It's in these statements:
```
temppath = '/scratch/' + getpass.getuser()
tempdir = '/scratch/' + getpass.getuser() + "/temp"
```

Re. "you should probably use `tempfile.TemporaryDirectory`" I'm not sure what that means, can you give an example of exactly what you want us to use here? We'll adopt whatever you think is best.

timj commented on DP0.2/13a_Image_Cutout_SciDemo.ipynb:
```diff
@@ -214,14 +215,20 @@
-    "del temppath"
+    "del temppath\n",
+    "\n",
+    "tempdir = '/scratch/' + getpass.getuser() + \"/temp\"\n",
```
I think you use `os.path` elsewhere in the notebook so can you also use it here: `tempdir = os.path.join(temppath, "temp")`? (you have already defined temppath above so no need to recalculate it here.

MelissaGraham replied:
I think this got fixed, resolving.

timj commented on DP0.2/13a_Image_Cutout_SciDemo.ipynb:
```diff
@@ -668,67 +666,84 @@
     "sq.raise_if_error()\n",
-    "with open(sodaCutout, 'bw') as f:\n",
-    "    f.write(response_stream.read())"
+    "mem = MemFileManager(len(cutout_bytes))\n",
```
Maybe use a function that takes the sq and returns the exposure (and use it above) rather than duplicating this somewhat esoteric afw code here?

MelissaGraham replied:
Done -- read_cutout_mem() now exists.

timj commented on DP0.2/13a_Image_Cutout_SciDemo.ipynb:
```diff
@@ -668,67 +666,84 @@
-    "# mem.setData(cutout_bytes, len(cutout_bytes))\n",
-    "# exposure = ExposureF(mem)\n",
-    "# plotImage(exposure)"
+    "sodaCutout = os.path.join(tempdir + '/cutout-circle.fits')\n",
```
If you are using os.path.join please use os.path.join -- the code here is not joining anything with `join`. Needs to be `os.path.join(tempdir, "cutout-circle.fits")`. If this file does not have to be kept around after the plotting it would be best to use `tempfile`.

```python
with tempfile.NamedTemporaryFile(dir=tempdir, suffix=".fits") as temp:
    response_stream = sq.execute_stream()
    sq.raise_if_error()
    temp.write(response_stream.read())
    temp.flush()
    plotImage(ExposureF(temp.name)
```
then the file will delete itself at the end of the cell.

timj replied:
Oh, and if you did it this way your tempdir could be the `/scratch/$USER` without the need for a new subdirectory.

MelissaGraham replied:
Oof yes, didn't catch those errors in the use of join. Also added a demo of saving to a temp file. Kept the subdirectory b/c we have to write and save the warped cutouts in a non-temporary way (I think?).

timj reviewed (approved):
Looks okay. I'm a bit concerned by the four places where MemFileManager is used where one function could do it instead (since that API is a little tricky). One suggestion for using NamedTemporaryFile.

MelissaGraham commented:
Made a few more changes:

* fixed to use os .join correctly
* made `read_cutout_mem` function
* added demo of save-to-temp-file

I think that captured all the suggestions thank you!
